db_client/functions/metadata.py

```diff
@@ -2,53 +2,87 @@
 
     NOTE: That the taxonomy is also validated. This is because the
     Taxonomy is stored in the database and can be mutated independently
     of the Family's metadata.
 
-    :param Session db: The db connection session.
-    :param Family family: The family to validate.
-    :raises TypeError: If the Taxonomy is invalid.
-    :raises ValidationError: If the Taxonomy values are invalid.
+    :param Session db: The Session to query.
+    :param str corpus_id: The corpus import ID to retrieve the taxonomy
+        for.
+    :param dict metadata: The family metadata to validate.
     :return Optional[MetadataValidationResult]: A list of errors or None
         if the metadata is valid.
     """
+    taxonomy = get_taxonomy_from_corpus(db, corpus_id)
+    if taxonomy is None:
+        raise TypeError("No taxonomy found for corpus")
+
+    # Make sure we only get the family taxonomy keys.
+    # TODO: When we move the family schema under _family we can consolidate these entity
+    # specific validation functions.
+    taxonomy = {
+        k: v for (k, v) in taxonomy.items() if k not in ["_document", "event_type"]
+    }
+    return _validate_metadata_against_taxonomy(taxonomy, metadata)
+
+
+def validate_document_metadata(
+    db, corpus_id: str, metadata
+) -> Optional[MetadataValidationErrors]:
+    """Validates the Document's metadata against its Corpus' Taxonomy.
+
+    NOTE: That the taxonomy is also validated. This is because the
+    Taxonomy is stored in the database and can be mutated independently
+    of the Family's metadata.
 
-    # Retrieve the Taxonomy and the Family's metadata
-    corpus_type, metadata = (
-        db.query(CorpusType, FamilyMetadata)
-        .join(Corpus, CorpusType.name == Corpus.corpus_type_name)
-        .join(FamilyCorpus, FamilyCorpus.corpus_import_id == Corpus.import_id)
-        .join(Family, family.import_id == FamilyCorpus.family_import_id)
-        .join(FamilyMetadata, FamilyMetadata.family_import_id == family.import_id)
-        .filter(Family.import_id == family.import_id)
-        .one()
-    )
-    taxonomy = corpus_type.valid_metadata
+    :param Session db: The Session to query.
+    :param str corpus_id: The corpus import ID to retrieve the taxonomy
+        for.
+    :param dict metadata: The document metadata to validate.
+    :return Optional[MetadataValidationResult]: A list of errors or None
+        if the metadata is valid.
+    """
+    taxonomy = get_taxonomy_from_corpus(db, corpus_id)
+    if taxonomy is None:
+        raise TypeError("No taxonomy found for corpus")
+
+    # Make sure we only get the document taxonomy keys.
+    # TODO: When we move the family schema under _family we can consolidate these entity
+    # specific validation functions.
+    taxonomy = get_entity_specific_taxonomy(taxonomy, "_document")
+    return _validate_metadata_against_taxonomy(taxonomy, metadata)
+
+
+def _validate_metadata_against_taxonomy(taxonomy, metadata):
```

Comment:
Would be good to get some type hints on here, and also make it a public function?

Reply:
Looking at this function again - maybe its worth just having as a test helper function - as it sort of hides the building of the taxonomy otherwise?

Reply:
This was more to avoid me having to write the same 7 lines of code in the family, document, and event validation functions